src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java

```diff
@@ -80,16 +74,65 @@ enum Signal {
   final Map<RepoRecordedInput, String> recordedInputValues = new TreeMap<>();
 
   SkyFunction.Environment signalForFreshEnv() throws InterruptedException {
-    signalQueue.put(Signal.RESTART);
-    return delegateEnvQueue.take();
+    // First unblock the Skyframe thread, then wait until it has something else to say. The only
+    // thing the Skyframe thread can do to this one is an interrupt so there is no need to do
+    // anything special in between.
+    rendezvousFromWorker(new Signal.Restart());
+    return rendezvousFromWorker(null);
+  }
+
+  SkyFunction.Environment rendezvousFromWorker(Signal signal) throws InterruptedException {
+    return (SkyFunction.Environment) rendezvous.exchange(signal);
+  }
+
+  Signal rendezvousFromHost(SkyFunction.Environment environment) throws InterruptedException {
+    return (Signal) rendezvous.exchange(environment);
+  }
+
+  Signal rendezvousUninterruptiblyFromHost(SkyFunction.Environment environment) {
+    while(true) {
+      try {
+        return (Signal) rendezvous.exchange(environment);
+      } catch (InterruptedException e) {
+      }
+    }
+  }
+
+  SkyFunction.Environment rendezvousUninterruptiblyFromWorker(Signal signal) {
+    while(true) {
+      try {
+        return (SkyFunction.Environment) rendezvous.exchange(signal);
+      } catch (InterruptedException e) {
+      }
+    }
+  }
+
+  public void join() {
+    Uninterruptibles.joinUninterruptibly(workerThread);
+    workerThread = null;
   }
 
   @Override
   public void close() {
-    var myWorkerFuture = workerFuture;
-    workerFuture = null;
-    if (myWorkerFuture != null) {
-      myWorkerFuture.cancel(true);
+    if (workerThread == null) {
+      return;
     }
+    workerThread.interrupt();
+
+    // Wait until the worker thread actually gets interrupted. Be resilient to cases where despite
```

Comment:
This goes against the comment on line 64 (ie. we're not supposed to block in close()).

Reply:
I realize that this is not supposed to block, but in my not-so-humble opinion it's much better to block a bit than to have unaccounted for state floating around for an unspecified amount of time. This is especially bad because the worker thread can have side effects visible outside of Bazel (since it can do I/O, write files, read files, etc., all from user code) and it sounds like a terrible idea to have that going on without the knowledge of the rest of Bazel.

That said, my sticking point is not that we block here, it's that we don't do anything with side effects outside of the worker thread after this method returns so if you have any clever ideas as to how to implement that without blocking, I'm all ears.

Reply:
> This is especially bad because the worker thread can have side effects visible outside of Bazel (since it can do I/O, write files, read files, etc., all from user code) and it sounds like a terrible idea to have that going on without the knowledge of the rest of Bazel.

Isn't that why we block in the host thread? The memory pressure listener thread just interrupts the worker thread, and it's the host thread's job to make sure the worker is finished before moving on. It's not necessary that the worker thread is terminated right after `close()`.